feat(Sequences): add random sequence generator

katestange opened the pull request:
closes #52

gwhitney commented:
I'll look at this as soon as I can if nobody else gets to it first. One tiny note: pull requests are generally by convention titled with what change they make, in the present tense, rather than by what "happened" in the past tense. Also, I find the "semantic commit messages" convention very helpful (see https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716). With those points in mind, I took the liberty of mildly renaming this PR.

katestange commented:
Thanks!

gwhitney commented:
OK, I've done the code review and have some suggested changes above. I'll continue the review/check-in process when those things are resolved. But definitely a nice addition to Numberscope!

gwhitney commented:
Oops, sorry, didn't realize I had to hit the "submit review" button for you to see the comments. They should be there now.

katestange commented on src/sequences/SequenceRandom.ts:
```diff
@@ -0,0 +1,83 @@
+	if ( !Number.isInteger( Number(this.settings.max) ) ) return new ValidationStatus(false, ["The max value must be an integer."]);
+
+	this.isValid = true;
+	this.name = "Random integers from " + String(Number(this.settings.min)) + " to " + String(Number(this.settings.max));
```
It fits ok in a browser (one line, within the provided box).  When we revisit UI it may need revisiting.

katestange commented on src/sequences/SequenceRandom.ts:
```diff
@@ -0,0 +1,83 @@
+	this.maximum = Number(this.settings.max);
+
+	console.log( "term", this.servedTerms[n] );
+	if ( isNaN(this.servedTerms[n])  ) {
```
It seems to work, yes.  But it requires a specific guarantee from the caching functionality that it won't recalculate terms ever.  So I've added an issue to add that as a test once we have a testing framework.

katestange commented:
Ok, ready for review again.

gwhitney commented on src/sequences/SequenceRandom.ts:
```diff
@@ -0,0 +1,77 @@
+
+		if ( !Number.isInteger( Number(this.settings.min) ) ) return new ValidationStatus(false, ["The min value must be an integer."]);
+		if ( !Number.isInteger( Number(this.settings.max) ) ) return new ValidationStatus(false, ["The max value must be an integer."]);
+
```
It's an unfortunate design flaw in the implementation of SequenceClassDefault that if either of the above cases are hit, `this.isValid` will be true: it's set to true in SequenceClassDefault.validate() and not reset to false in either of these cases, unlike in the cases where 'min' or 'max' is undefined. I'll create an issue for this design flaw, but in the meantime, the expedient thing to do for this PR would be to set this.isValid to false immediately after checking the return from super.validate() like at line 47. Then you can go through the bad conditions, with the isValid flag safely false if any of them are hit (i.e. you can then remove line 49) , and then you already set it back to true below once you've gotten through all the bad conditions, so it will be OK if there are no bad conditions. Hope I am making sense.

gwhitney commented on src/sequences/SequenceRandom.ts:
```diff
@@ -0,0 +1,77 @@
+	calculate(n: number) {
+		this.minimum = Number(this.settings.min);
+		this.maximum = Number(this.settings.max);
+
```
You should set `this.minimum` and `this.maximum` once in validate(), where in fact you are already doing the conversion to Number anyway, rather than re-doing the conversion on every call to calculate().

katestange replied:
Makes sense. General question:  where does one do general setup for a more general sequence?  validate() seems too special-purpose, at least based on its name, but sequences don't have a general setup() function that would only get called once.  Should that be an issue to add?

gwhitney replied:
There is already an initialize() method in SequenceInterface that is guaranteed to be called once per sequence. But it is called _after_ validate(), and by default checks that validation succeeded. But in the case of your two parameters, they already have to be parsed in validate(), so there is no reason not to set the object properties right in validate() as you have done. On the other hand, I agree, if you had to say pre-compute some big matrix or something like that one time before calculate() (or getElement(), in case of an uncached Sequence) was called, you would do that in initialize() rather than validate().

gwhitney commented on src/sequences/SequenceRandom.ts:
```diff
@@ -0,0 +1,77 @@
+		'min',
+		'number',
+		'Minimum value attainable',
+		false,
```
This is the argument that specifies whether the parameter `min` is required. Looking at validate() below, clearly it is required. Hence this should be true, and the same goes for `max` below. I will also file an issue that if a parameter has been marked as required, the Sequence writer should not also have to check it manually in validate(). (Right now this parameter and the code in validate() are partially redundant.)

gwhitney reviewed (commented):
My apologies that my first review wasn't thorough enough. On closer inspection, uncovered a couple more concerns, most of which actually stem from minor design flaws in the Sequence hierarchy that I will file issues about. But they do need fixing here for now for this to go in, since I don't think it makes sense to wait for them to be fixed in general.

katestange commented:
Just pushed again.  Fixed the above, also caught a naming error in line 42, added something to suppress a lint error because n is not used in calculate() (line 65).

gwhitney commented:
(Just got rid of trailing whitespace, mixed tabs/spaces, and a line with only whitespace. All of these things are highlighted by "git diff main", which you should always do before submitting a PR.)

gwhitney commented:
Since the integer 0 is a "falsy" value, the default for the min of the range was not showing up pre-filled in the sequence parameter dialog. Fixed this by changing it to the one-character string `'0'` which is a "truthy" value.
